### DISCUSSION TITLE
MAS vs. Synapse Homeserver OIDC Providers

### DISCUSSION THREAD

**Discussion by Viajaz** [element-hq/matrix-authentication-service#3485] [Q&A]
I'm looking for some information on why a Synapse homeserver operator would want to use MAS (MSC3861) if they already have [OpenID Connect Provider(s)](https://element-hq.github.io/synapse/latest/openid.html) configured on their homeserver(s) (`oidc_providers`) for their desired Identity Provider(s).

**Comment by ktims** [comment]
I'm also wondering this. It seems like the solution here is backwards. There really needs to be a document explaining why it is better to have an additional service than to have Synapse just integrate with an existing IdP _as already exists_ and why this functionality is being deprecated. It's a bit frustrating to set up a whole OIDC provider just to more or less proxy requests to the OIDC provider I actually want to use.

**Comment by sandhose** [comment]
MAS isn't an IDP, it's a rewrite auth Synapse's internal auth system. It's highly tied to Synapse, as it needs to manage Matrix devices, users, etc.
OIDC/OAuth 2.0 are **authorisation** frameworks, and don't cover (well) actual session lifetime, what to do when they end; it also doesn't tell you what to do with dynamic client registrations (just how you'd do them, not what you need to enforce/accept/reject)

I think we did the mistake of naming this new API 'OIDC native', whilst it really the 'next gen auth API for Matrix, based on OAuth 2.0/OIDC'.

I encourage you watching my talk at the Matrix conference, the second half talks about MAS itself: https://youtu.be/wOW8keNafdE?t=919

**Comment by rbi** [comment]
When it is highly tied to Synapse anyway, why is extracting an arbitrary feature "authentication" into a separate service a good thing? Couldn't it just be integrated into Synapse itself?

I wanted to try out Element X with my Synapse home server which doesn't work just because I use the "old" oidc_providers feature of Synapse. I was about to install MAS in my system. I came to the point where I should setup another sub domain just for MAS but should redirect some arbitrary paths of the Synapse domain to MAS as well for keeping "legacy auth" working. That's where I decided that just for trying out Element X it is not worth the hassle and the huge additional infrastructure complexity.
